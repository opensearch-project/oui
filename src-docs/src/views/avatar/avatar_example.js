/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';

import { renderToHtml } from '../../services';

import { GuideSectionTypes } from '../../components';

import { OuiAvatar, OuiCode } from '../../../../src/components';
import avatarConfig from './playground';

import Avatar from './avatar';
const avatarSource = require('!!raw-loader!./avatar');
const avatarHtml = renderToHtml(Avatar);
const avatarSnippet = [
  `<OuiAvatar name="Raphael" />
`,
  `<OuiAvatar size="s" name="Cat" imageUrl="https://source.unsplash.com/64x64/?cat" />
`,
  '<OuiAvatar name="Leonardo" color="#BD10E0" />',
];

import AvatarInitials from './avatar_initials';
const avatarInitialsSource = require('!!raw-loader!./avatar_initials');
const avatarInitialsHtml = renderToHtml(AvatarInitials);
const avatarInitialsSnippet = [
  `<OuiAvatar name="OpenSearch Dashboards" initialsLength={2}/>
`,
  `<OuiAvatar name="Undefined" initials="?" />
`,
  `<OuiAvatar name="Engineering User" initials="En" initialsLength={2} />
`,
];

import AvatarTypes from './avatar_type';
const avatarTypesSource = require('!!raw-loader!./avatar_type');
const avatarTypesHtml = renderToHtml(AvatarTypes);
const avatarTypesSnippet = [
  `<OuiAvatar type="space" name="Engineering Space" />
`,
];

import AvatarIcons from './avatar_icon';
const avatarIconsSource = require('!!raw-loader!./avatar_icon');
const avatarIconsHtml = renderToHtml(AvatarIcons);
const avatarIconsSnippet = [
  `<OuiAvatar name="Management" iconType="managementApp" />
`,
  `<OuiAvatar name="Management" iconType="managementApp" color="#FAFBFD" iconColor={null} />
`,
  `<OuiAvatar name="Management" iconType="managementApp" iconSize="l" />
`,
];

import AvatarDisabled from './avatar_disabled';
const avatarDisabledSource = require('!!raw-loader!./avatar_disabled');
const avatarDisabledHtml = renderToHtml(AvatarDisabled);
const avatarDisabledSnippet = [
  `<OuiAvatar isDisabled={true} name="Avatar" />
`,
];

export const AvatarExample = {
  title: 'Avatar',
  sections: [
    {
      source: [
        {
          type: GuideSectionTypes.JS,
          code: avatarSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: avatarHtml,
        },
      ],
      text: (
        <div>
          <p>
            The <strong>OuiAvatar</strong> component typically creates a user
            icon. It will accept <OuiCode>name</OuiCode> (required) and{' '}
            <OuiCode>image</OuiCode> props and will configure the display and
            accessibility as needed. By default, the background colors come from
            the set of colors used for visualizations. Otherwise you can pass a
            hex value to the <OuiCode>color</OuiCode> prop.
          </p>
        </div>
      ),
      props: { OuiAvatar },
      snippet: avatarSnippet,
      demo: <Avatar />,
    },
    {
      title: 'Initials',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: avatarInitialsSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: avatarInitialsHtml,
        },
      ],
      text: (
        <div>
          <p>
            The initials displayed in the avatar try to be smart based on the
            name prop. If the name contains spaces, it will display the first
            character of each word,{' '}
            <strong>always maxing out at 2 characters</strong>. You can
            customize this by passing a combination of{' '}
            <OuiCode>initialsLength</OuiCode> and/or <OuiCode>initials</OuiCode>{' '}
            props. However, the avatar will still always max out at 2
            characters.
          </p>
        </div>
      ),
      snippet: avatarInitialsSnippet,
      demo: <AvatarInitials />,
      props: { OuiAvatar },
    },
    {
      title: 'Types',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: avatarTypesSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: avatarTypesHtml,
        },
      ],
      text: (
        <div>
          <p>
            The avatar <OuiCode>type</OuiCode>, which primarily defines the
            shape, is keyworded and can be{' '}
            <OuiCode language="js">&quot;user&quot;</OuiCode> (default) or{' '}
            <OuiCode language="js">&quot;space&quot;</OuiCode> (for workspaces).
          </p>
        </div>
      ),
      snippet: avatarTypesSnippet,
      demo: <AvatarTypes />,
      props: { OuiAvatar },
    },
    {
      title: 'Icons',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: avatarIconsSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: avatarIconsHtml,
        },
      ],
      text: (
        <div>
          <p>
            Icons can also be displayed instead of initials or images. When
            simply passing an <OuiCode>iconType</OuiCode>, it will both size and
            color appropriately based on the other <strong>OuiAvatar</strong>{' '}
            props. To customize these specifically, pass{' '}
            <OuiCode>iconSize</OuiCode> and <OuiCode>iconColor</OuiCode>.
          </p>
          <p>
            If your icon has multiples or custom colors like a logo, you can
            keep the default <OuiCode>iconColor</OuiCode> by passing{' '}
            <OuiCode>null</OuiCode>. Otherwise it will get the appropriate
            contrast acceptable variant. Just ensure that you also are providing
            an accesible background color to match that of the icon&apos;s
            color.
          </p>
        </div>
      ),
      snippet: avatarIconsSnippet,
      demo: <AvatarIcons />,
      props: { OuiAvatar },
    },
    {
      title: 'Disabled',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: avatarDisabledSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: avatarDisabledHtml,
        },
      ],
      text: (
        <div>
          <p>
            While <strong>OuiAvatar</strong> doesn&apos;t accept any interactive
            behaviors itself, you can create a visually presented disabled
            avatar by adding <OuiCode>isDisabled</OuiCode> when placed within a
            disabled element.
          </p>
        </div>
      ),
      snippet: avatarDisabledSnippet,
      demo: <AvatarDisabled />,
      props: { OuiAvatar },
    },
  ],
  playground: avatarConfig,
};
