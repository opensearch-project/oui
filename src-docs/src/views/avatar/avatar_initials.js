/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';

import { OuiAvatar, OuiTitle, OuiSpacer } from '../../../../src/components';

export default () => (
  <div>
    <OuiTitle size="xs">
      <h3>Single vs multi-word</h3>
    </OuiTitle>
    <OuiSpacer />
    <OuiAvatar name="Single" />
    &emsp;
    <OuiAvatar name="Two Words" />
    &emsp;
    <OuiAvatar name="More Than Two Words" />
    &emsp;
    <OuiAvatar name="lowercase words" />
    <OuiSpacer />
    <OuiTitle size="xs">
      <h4>Custom</h4>
    </OuiTitle>
    <OuiSpacer />
    <OuiAvatar name="Dashboards" initialsLength={2} />
    &emsp;
    <OuiAvatar name="Leonardo Dude" initialsLength={1} />
    &emsp;
    <OuiAvatar name="Not provided" initials="?" />
    &emsp;
    <OuiAvatar name="Engineering User" initials="En" initialsLength={2} />
  </div>
);
