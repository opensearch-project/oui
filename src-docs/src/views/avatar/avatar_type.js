/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';

import { OuiAvatar, OuiTitle, OuiSpacer } from '../../../../src/components';

export default () => (
  <div>
    <OuiTitle size="xs">
      <h3>Spaces</h3>
    </OuiTitle>
    <OuiSpacer />
    <OuiAvatar size="s" type="space" name="Dashboards" />
    &emsp;
    <OuiAvatar type="space" name="Leonardo Space" />
    &emsp;
    <OuiAvatar size="l" type="space" name="Default" />
    &emsp;
    <OuiAvatar size="xl" type="space" name="Engineering Space" />
  </div>
);
