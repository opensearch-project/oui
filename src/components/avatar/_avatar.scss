/*!
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

.ouiAvatar {
  flex-shrink: 0; // Ensures it never scales down below its intended size
  display: inline-flex;
  justify-content: center;
  align-items: center;
  background-size: cover;
  text-align: center;
  vertical-align: middle;
  overflow-x: hidden;
  font-weight: $ouiFontWeightMedium; // Explicitly state so it doesn't get overridden by inheritance
}

.ouiAvatar--user {
  @include innerBorder('dark', 50%, .05);

  border-radius: 50%;
}

.ouiAvatar--space {
  @include innerBorder('dark', $ouiBorderRadius, .05);

  border-radius: $ouiBorderRadius;
}

.ouiAvatar-isDisabled {
  cursor: not-allowed;
  filter: grayscale(100%);
}

.ouiAvatar--plain {
  background-color: $ouiColorEmptyShade;
}

// Modifiers for sizing.
$avatarSizing: (
  s: (
    size: $ouiSizeL,
    font-size: $ouiSizeM
  ),
  m: (
    size: $ouiSizeXL,
    font-size: $ouiSize * .9
  ),
  l: (
    size: $ouiSizeXXL,
    font-size: $ouiSizeL * .8
  ),
  xl: (
    size: ($ouiSize * 4),
    font-size: $ouiSizeXL * .8
  ),
);

@each $size, $map in $avatarSizing {
  .ouiAvatar--#{$size} {
    @include size(map-get($map, 'size'));

    line-height: map-get($map, 'size');
    font-size: map-get($map, 'font-size');
  }
}
