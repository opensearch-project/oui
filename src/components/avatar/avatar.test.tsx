/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render } from '@testing-library/react';
import { requiredProps } from '../../test/required_props';

import { OuiAvatar, SIZES } from './avatar';

describe('OuiAvatar', () => {
  test('is rendered', () => {
    const { container } = render(<OuiAvatar name="name" {...requiredProps} />);

    expect(container).toMatchSnapshot();
  });

  test('allows a name composed entirely of whitespace', () => {
    const { container } = render(<OuiAvatar name="  " {...requiredProps} />);

    expect(container).toMatchSnapshot();
  });

  describe('props', () => {
    describe('imageUrl', () => {
      it('is rendered', () => {
        const { container } = render(
          <OuiAvatar name="name" imageUrl="image url" />
        );

        expect(container).toMatchSnapshot();
      });
    });

    describe('iconType', () => {
      it('is rendered', () => {
        const { container } = render(<OuiAvatar name="name" iconType="bolt" />);

        expect(container).toMatchSnapshot();
      });

      it('and iconSize is rendered', () => {
        const { container } = render(
          <OuiAvatar name="name" iconType="bolt" iconSize="xl" />
        );

        expect(container).toMatchSnapshot();
      });

      it('and iconColor is rendered', () => {
        const { container } = render(
          <OuiAvatar name="name" iconType="bolt" iconColor="primary" />
        );

        expect(container).toMatchSnapshot();
      });

      it('and iconColor as null is rendered', () => {
        const { container } = render(
          <OuiAvatar name="name" iconType="bolt" iconColor={null} />
        );

        expect(container).toMatchSnapshot();
      });
    });

    describe('size', () => {
      SIZES.forEach((size) => {
        it(`${size} is rendered`, () => {
          const { container } = render(<OuiAvatar name="name" size={size} />);

          expect(container).toMatchSnapshot();
        });
      });
    });

    describe('initials', () => {
      it('is rendered', () => {
        const { container } = render(<OuiAvatar name="name" initials="lo" />);

        expect(container).toMatchSnapshot();
      });
    });

    describe('initialsLength', () => {
      it('is rendered', () => {
        const { container } = render(
          <OuiAvatar name="name" initialsLength={2} />
        );

        expect(container).toMatchSnapshot();
      });
    });

    describe('type', () => {
      it('is rendered', () => {
        const { container } = render(<OuiAvatar name="name" type="space" />);

        expect(container).toMatchSnapshot();
      });
    });

    describe('color', () => {
      it('as string is rendered', () => {
        const { container } = render(<OuiAvatar name="name" color="#000" />);

        expect(container).toMatchSnapshot();
      });

      it('as null is rendered', () => {
        const { container } = render(<OuiAvatar name="name" color={null} />);

        expect(container).toMatchSnapshot();
      });

      it('as plain is rendered', () => {
        const { container } = render(<OuiAvatar name="name" color="plain" />);

        expect(container).toMatchSnapshot();
      });
    });

    describe('isDisabled', () => {
      it('is rendered', () => {
        const { container } = render(
          <OuiAvatar name="name" isDisabled={true} />
        );

        expect(container).toMatchSnapshot();
      });
    });
  });

  test('should throw error if color is not a hex', () => {
    const renderComponent = () =>
      render(<OuiAvatar name="name" color="rgba(0,0,0,0)" />);

    expect(renderComponent).toThrowErrorMatchingSnapshot();
  });
});
