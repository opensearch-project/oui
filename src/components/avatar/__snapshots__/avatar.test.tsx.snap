// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiAvatar allows a name composed entirely of whitespace 1`] = `
<div>
  <div
    aria-label="aria-label"
    class="ouiAvatar ouiAvatar--m ouiAvatar--user testClass1 testClass2"
    data-test-subj="test subject string"
    style="background-color: rgb(238, 120, 157); color: rgb(0, 0, 0);"
    title="  "
  >
    <span
      aria-hidden="true"
    >
        
    </span>
  </div>
</div>
`;

exports[`OuiAvatar is rendered 1`] = `
<div>
  <div
    aria-label="aria-label"
    class="ouiAvatar ouiAvatar--m ouiAvatar--user testClass1 testClass2"
    data-test-subj="test subject string"
    style="background-color: rgb(228, 166, 199); color: rgb(0, 0, 0);"
    title="name"
  >
    <span
      aria-hidden="true"
    >
      n
    </span>
  </div>
</div>
`;

exports[`OuiAvatar props color as null is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--m ouiAvatar--user"
    title="name"
  >
    <span
      aria-hidden="true"
    >
      n
    </span>
  </div>
</div>
`;

exports[`OuiAvatar props color as plain is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--m ouiAvatar--user ouiAvatar--plain"
    title="name"
  >
    <span
      aria-hidden="true"
    >
      n
    </span>
  </div>
</div>
`;

exports[`OuiAvatar props color as string is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--m ouiAvatar--user"
    style="background-color: rgb(0, 0, 0); color: rgb(255, 255, 255);"
    title="name"
  >
    <span
      aria-hidden="true"
    >
      n
    </span>
  </div>
</div>
`;

exports[`OuiAvatar props iconType and iconColor as null is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--m ouiAvatar--user"
    style="background-color: rgb(228, 166, 199); color: rgb(0, 0, 0);"
    title="name"
  >
    <span
      aria-label="name"
      class="ouiAvatar__icon"
      data-ouiicon-type="bolt"
    />
  </div>
</div>
`;

exports[`OuiAvatar props iconType and iconColor is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--m ouiAvatar--user"
    style="background-color: rgb(228, 166, 199); color: rgb(0, 0, 0);"
    title="name"
  >
    <span
      aria-label="name"
      class="ouiAvatar__icon"
      color="primary"
      data-ouiicon-type="bolt"
    />
  </div>
</div>
`;

exports[`OuiAvatar props iconType and iconSize is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--m ouiAvatar--user"
    style="background-color: rgb(228, 166, 199); color: rgb(0, 0, 0);"
    title="name"
  >
    <span
      aria-label="name"
      class="ouiAvatar__icon"
      color="#000000"
      data-ouiicon-type="bolt"
    />
  </div>
</div>
`;

exports[`OuiAvatar props iconType is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--m ouiAvatar--user"
    style="background-color: rgb(228, 166, 199); color: rgb(0, 0, 0);"
    title="name"
  >
    <span
      aria-label="name"
      class="ouiAvatar__icon"
      color="#000000"
      data-ouiicon-type="bolt"
    />
  </div>
</div>
`;

exports[`OuiAvatar props imageUrl is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--m ouiAvatar--user"
    style="background-color: rgb(228, 166, 199); color: rgb(0, 0, 0);"
    title="name"
  />
</div>
`;

exports[`OuiAvatar props initials is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--m ouiAvatar--user"
    style="background-color: rgb(228, 166, 199); color: rgb(0, 0, 0);"
    title="name"
  >
    <span
      aria-hidden="true"
    >
      l
    </span>
  </div>
</div>
`;

exports[`OuiAvatar props initialsLength is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--m ouiAvatar--user"
    style="background-color: rgb(228, 166, 199); color: rgb(0, 0, 0);"
    title="name"
  >
    <span
      aria-hidden="true"
    >
      na
    </span>
  </div>
</div>
`;

exports[`OuiAvatar props isDisabled is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--m ouiAvatar--user ouiAvatar-isDisabled"
    style="background-color: rgb(228, 166, 199); color: rgb(0, 0, 0);"
    title="name"
  >
    <span
      aria-hidden="true"
    >
      n
    </span>
  </div>
</div>
`;

exports[`OuiAvatar props size l is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--l ouiAvatar--user"
    style="background-color: rgb(228, 166, 199); color: rgb(0, 0, 0);"
    title="name"
  >
    <span
      aria-hidden="true"
    >
      n
    </span>
  </div>
</div>
`;

exports[`OuiAvatar props size m is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--m ouiAvatar--user"
    style="background-color: rgb(228, 166, 199); color: rgb(0, 0, 0);"
    title="name"
  >
    <span
      aria-hidden="true"
    >
      n
    </span>
  </div>
</div>
`;

exports[`OuiAvatar props size s is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--s ouiAvatar--user"
    style="background-color: rgb(228, 166, 199); color: rgb(0, 0, 0);"
    title="name"
  >
    <span
      aria-hidden="true"
    >
      n
    </span>
  </div>
</div>
`;

exports[`OuiAvatar props size xl is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--xl ouiAvatar--user"
    style="background-color: rgb(228, 166, 199); color: rgb(0, 0, 0);"
    title="name"
  >
    <span
      aria-hidden="true"
    >
      n
    </span>
  </div>
</div>
`;

exports[`OuiAvatar props type is rendered 1`] = `
<div>
  <div
    aria-label="name"
    class="ouiAvatar ouiAvatar--m ouiAvatar--space"
    style="background-color: rgb(228, 166, 199); color: rgb(0, 0, 0);"
    title="name"
  >
    <span
      aria-hidden="true"
    >
      n
    </span>
  </div>
</div>
`;

exports[`OuiAvatar should throw error if color is not a hex 1`] = `"Error: Uncaught [Error: OuiAvatar needs to pass a valid color. This can either be a three or six character hex value]"`;
